src/System.Linq/tests/ConcatTests.cs

```diff
@@ -206,6 +206,16 @@ public void CountOfConcatIteratorShouldThrowExceptionOnIntegerOverflow()
             Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
+
+            // This applies to ToArray() and ToList() as well, which try to preallocate the exact size
+            // needed if all inputs are ICollections.
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
```

Comment:
If we turned a situation that would have worked into one that now throws, we need to fix that, and we shouldn't double-down on it and introduce additional such cases.  It's one thing to say that in extreme circumstances we may throw an OverflowException instead of an OutOfMemoryException (though even that is something we should carefully consider), but it's another entirely to say that we're taking something that returned successfully and instead making it throw... the latter is bad.